pkg/reconciler/autoscaling/versionwrapper/controller.go

```diff
@@ -0,0 +1,68 @@
+package versionwrapper
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/blang/semver/v4"
+	"k8s.io/client-go/discovery"
+	kubeclient "knative.dev/pkg/client/injection/kube/client"
+	"knative.dev/pkg/configmap"
+	"knative.dev/pkg/controller"
+	"knative.dev/serving/pkg/reconciler/autoscaling/hpa"
+	hpav2beta2 "knative.dev/serving/pkg/reconciler/autoscaling/hpav2beta2"
+)
+
+func NewController(
+	ctx context.Context,
+	cmw configmap.Watcher,
+) *controller.Impl {
+	kc := kubeclient.Get(ctx)
+	// Starting from 4.10 (1.24.0) we can use the new API version (autoscaling/v2) of HorizontalPodAutoscaler
+	// As we also need to support 4.8 we also need provide the controller using the old API version (autoscaling/v2beta2)
+	if err := CheckMinimumVersion(kc.Discovery(), "1.24.0"); err == nil {
```

Comment:
Not sure if we avoid the imports of v2beta2 and vice versa. Since hpav2beta2 brings in `hpainformer "knative.dev/pkg/client/injection/kube/informers/autoscaling/v2beta2/horizontalpodautoscaler"` .

Reply:
Yeah not sure about that. But I'm not sure how we could fix it, if this does not work. As the init functions are in the generated client code. I'm going to try it on my OCP cluster.